compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1778,8 +1778,9 @@ impl<'a> Parser<'a> {
             let (pat, ty) = if is_name_required || this.is_named_param() {
                 debug!("parse_param_general parse_pat (is_name_required:{})", is_name_required);
 
-                let pat = this.parse_fn_param_pat()?;
-                if let Err(mut err) = this.expect(&token::Colon) {
+                let (pat, colon) = this.parse_fn_param_pat_colon()?;
+                if !colon {
+                    let mut err = this.expected_one_of_not_found(&[], &[token::Colon]).unwrap_err();
```

Comment:
Colon should already be in expected tokens, so simple `this.unexpected()` should work too.